File: src/midi/SmfWriter.js
import { mkdir, writeFile } from 'node:fs/promises'
import { dirname } from 'node:path'

const PPQ = 480

export function arrangementToSmf(arrangement, transport) {
  const tempoMap = (transport?.tempoMap?.length ? transport.tempoMap : [{ beat: 0, bpm: 120 }])
    .slice()
    .sort((a, b) => a.beat - b.beat)

  const byNode = new Map()
  for (const clip of (arrangement?.midiClips ?? [])) {
    if (!byNode.has(clip.targetNodeId)) byNode.set(clip.targetNodeId, [])
    byNode.get(clip.targetNodeId).push(clip)
  }

  const tracks = [buildTempoTrack(tempoMap)]
  for (const [nodeId, clips] of byNode) tracks.push(buildNoteTrack(nodeId, clips))

  return buildSmf(tracks)
}

export async function writeSmf(filePath, arrangement, transport) {
  const buffer = arrangementToSmf(arrangement, transport)
  await mkdir(dirname(filePath), { recursive: true })
  await writeFile(filePath, buffer)
  const clips = arrangement?.midiClips ?? []
  const tracks = new Set(clips.map(c => c.targetNodeId)).size
  return { filePath, bytes: buffer.length, tracks, clips: clips.length }
}

function buildTempoTrack(tempoMap) {
  const events = [trackNameEvent(0, 'Tempo')]
  for (const { beat, bpm } of tempoMap) {
    const usPerBeat = Math.round(60_000_000 / bpm)
    events.push({
      tick: beatToTick(beat),
      bytes: [0xFF, 0x51, 0x03,
        (usPerBeat >> 16) & 0xFF, (usPerBeat >> 8) & 0xFF, usPerBeat & 0xFF]
    })
  }
  events.push(endOfTrack(events[events.length - 1].tick))
  return eventsToTrack(events)
}

function buildNoteTrack(nodeId, clips) {
  const label = nodeId.includes('/') ? nodeId.split('/').pop() : nodeId
  const events = [trackNameEvent(0, label)]

  for (const clip of clips) {
    for (const note of clip.notes) {
      const absStart = clip.startBeat + note.startBeat
      const absEnd = absStart + note.durationBeats
      const ch = note.channel & 0x0F
      events.push({ tick: beatToTick(absStart), bytes: [0x90 | ch, note.pitch, note.velocity] })
      events.push({ tick: beatToTick(absEnd),   bytes: [0x80 | ch, note.pitch, 0] })
    }
  }

  events.sort((a, b) => {
    if (a.tick !== b.tick) return a.tick - b.tick
    const aOff = (a.bytes[0] & 0xF0) === 0x80
    const bOff = (b.bytes[0] & 0xF0) === 0x80
    return aOff === bOff ? 0 : aOff ? -1 : 1
  })

  events.push(endOfTrack(events[events.length - 1].tick))
  return eventsToTrack(events)
}

function eventsToTrack(events) {
  const data = []
  let currentTick = 0
  for (const event of events) {
    const delta = Math.max(0, event.tick - currentTick)
    currentTick = event.tick
    data.push(...varLen(delta), ...event.bytes)
  }
  const chunk = Buffer.alloc(8 + data.length)
  chunk.write('MTrk', 0, 'ascii')
  chunk.writeUInt32BE(data.length, 4)
  for (let i = 0; i < data.length; i++) chunk[8 + i] = data[i]
  return chunk
}

function buildSmf(trackBuffers) {
  const header = Buffer.alloc(14)
  header.write('MThd', 0, 'ascii')
  header.writeUInt32BE(6, 4)
  header.writeUInt16BE(trackBuffers.length === 1 ? 0 : 1, 8)
  header.writeUInt16BE(trackBuffers.length, 10)
  header.writeUInt16BE(PPQ, 12)
  return Buffer.concat([header, ...trackBuffers])
}

export async function writeSmfFromEvents(filePath, events, nodeLabels = {}, transport = null) {
  const tempoMap = (transport?.tempoMap?.length ? transport.tempoMap : [{ beat: 0, bpm: 120 }])
    .slice().sort((a, b) => a.beat - b.beat)

  const byNode = new Map()
  for (const event of (events ?? [])) {
    if (!byNode.has(event.nodeId)) byNode.set(event.nodeId, [])
    byNode.get(event.nodeId).push(event)
  }

  const tracks = [buildTempoTrack(tempoMap)]
  for (const [nodeId, nodeEvents] of byNode) {
    const label = nodeLabels[nodeId] ?? nodeId.split('/').pop()
    tracks.push(buildRawEventTrack(label, nodeEvents))
  }

  const buffer = buildSmf(tracks)
  await mkdir(dirname(filePath), { recursive: true })
  await writeFile(filePath, buffer)
  return { filePath, bytes: buffer.length, tracks: byNode.size, events: (events ?? []).length }
}

function buildRawEventTrack(label, events) {
  const sorted = events
    .filter(e => e.status >= 0x80 && e.status <= 0xEF)
    .map(e => ({ tick: beatToTick(e.beatPosition ?? 0), bytes: [e.status, e.data1 ?? 0, e.data2 ?? 0] }))
    .sort((a, b) => a.tick - b.tick)

  const trackEvents = [trackNameEvent(0, label), ...sorted, endOfTrack(sorted[sorted.length - 1]?.tick ?? 0)]
  return eventsToTrack(trackEvents)
}

function trackNameEvent(tick, name) {
  const bytes = Array.from(name.slice(0, 127)).map(c => c.charCodeAt(0))
  return { tick, bytes: [0xFF, 0x03, bytes.length, ...bytes] }
}

function endOfTrack(tick) {
  return { tick, bytes: [0xFF, 0x2F, 0x00] }
}

function beatToTick(beat) {
  return Math.round(beat * PPQ)
}

function varLen(value) {
  if (value < 0x80) return [value]
  const bytes = []
  let v = value
  while (v > 0) { bytes.unshift(v & 0x7F); v >>>= 7 }
  for (let i = 0; i < bytes.length - 1; i++) bytes[i] |= 0x80
  return bytes
}

File: native/src/SmfWriter.cpp
#include "transmission/SmfWriter.h"

#include <algorithm>
#include <cmath>
#include <filesystem>
#include <fstream>
#include <map>
#include <vector>

namespace transmission {
namespace {

constexpr std::uint16_t PPQ = 480;

void writeU32BE(std::vector<std::uint8_t>& buf, std::uint32_t v) {
    buf.push_back(static_cast<std::uint8_t>((v >> 24) & 0xFF));
    buf.push_back(static_cast<std::uint8_t>((v >> 16) & 0xFF));
    buf.push_back(static_cast<std::uint8_t>((v >> 8) & 0xFF));
    buf.push_back(static_cast<std::uint8_t>(v & 0xFF));
}

void writeU16BE(std::vector<std::uint8_t>& buf, std::uint16_t v) {
    buf.push_back(static_cast<std::uint8_t>((v >> 8) & 0xFF));
    buf.push_back(static_cast<std::uint8_t>(v & 0xFF));
}

void writeVarLen(std::vector<std::uint8_t>& buf, std::uint32_t v) {
    if (v < 0x80) { buf.push_back(static_cast<std::uint8_t>(v)); return; }
    std::uint8_t bytes[4]{};
    int n = 0;
    while (v > 0) { bytes[n++] = static_cast<std::uint8_t>(v & 0x7F); v >>= 7; }
    for (int i = n - 1; i > 0; --i) buf.push_back(bytes[i] | 0x80);
    buf.push_back(bytes[0]);
}

struct RawEvent {
    std::uint32_t tick = 0;
    std::vector<std::uint8_t> bytes;
};

std::uint32_t beatToTick(double beat) {
    return static_cast<std::uint32_t>(std::round(beat * PPQ));
}

std::vector<std::uint8_t> eventsToTrack(std::vector<RawEvent> events) {
    std::vector<std::uint8_t> data;
    std::uint32_t currentTick = 0;
    for (const auto& ev : events) {
        const auto delta = ev.tick >= currentTick ? ev.tick - currentTick : 0U;
        currentTick = ev.tick;
        writeVarLen(data, delta);
        data.insert(data.end(), ev.bytes.begin(), ev.bytes.end());
    }
    std::vector<std::uint8_t> chunk;
    chunk.push_back('M'); chunk.push_back('T');
    chunk.push_back('r'); chunk.push_back('k');
    writeU32BE(chunk, static_cast<std::uint32_t>(data.size()));
    chunk.insert(chunk.end(), data.begin(), data.end());
    return chunk;
}

std::vector<std::uint8_t> buildTempoTrack(double bpm) {
    const auto usPerBeat = static_cast<std::uint32_t>(
        std::round(60'000'000.0 / (bpm > 0.0 ? bpm : 120.0)));
    std::vector<RawEvent> events;
    events.push_back({0, {0xFF, 0x03, 0x05, 'T', 'e', 'm', 'p', 'o'}});
    events.push_back({0, {0xFF, 0x51, 0x03,
        static_cast<std::uint8_t>((usPerBeat >> 16) & 0xFF),
        static_cast<std::uint8_t>((usPerBeat >> 8) & 0xFF),
        static_cast<std::uint8_t>(usPerBeat & 0xFF)}});
    events.push_back({0, {0xFF, 0x2F, 0x00}});
    return eventsToTrack(std::move(events));
}

std::vector<std::uint8_t> buildNoteTrack(
    const std::string& nodeId,
    const std::vector<const UiProjectMidiClip*>& clips)
{
    const auto slash = nodeId.rfind('/');
    const auto label = slash != std::string::npos
        ? nodeId.substr(slash + 1) : nodeId;
    const auto nameLen = static_cast<std::uint8_t>(
        std::min(label.size(), std::size_t{127}));
    std::vector<RawEvent> events;
    std::vector<std::uint8_t> nameBytes = {0xFF, 0x03, nameLen};
    nameBytes.insert(nameBytes.end(), label.begin(), label.begin() + nameLen);
    events.push_back({0, std::move(nameBytes)});

    for (const auto* clip : clips) {
        for (const auto& note : clip->notes) {
            const double absStart = clip->startBeat + note.startBeat;
            const double absEnd = absStart + note.durationBeats;
            const auto ch = static_cast<std::uint8_t>(note.channel & 0x0F);
            events.push_back({beatToTick(absStart),
                {static_cast<std::uint8_t>(0x90 | ch), note.pitch, note.velocity}});
            events.push_back({beatToTick(absEnd),
                {static_cast<std::uint8_t>(0x80 | ch), note.pitch, 0x00}});
        }
    }

    // Sort by tick; note-offs before note-ons at the same tick.
    std::stable_sort(events.begin() + 1, events.end(),
        [](const RawEvent& a, const RawEvent& b) {
            if (a.tick != b.tick) return a.tick < b.tick;
            const bool aOff = !a.bytes.empty() && (a.bytes[0] & 0xF0) == 0x80;
            const bool bOff = !b.bytes.empty() && (b.bytes[0] & 0xF0) == 0x80;
            return aOff && !bOff;
        });

    const auto lastTick = events.empty() ? 0U : events.back().tick;
    events.push_back({lastTick, {0xFF, 0x2F, 0x00}});
    return eventsToTrack(std::move(events));
}

} // namespace

std::vector<std::uint8_t> arrangementToSmf(
    const std::vector<UiProjectMidiClip>& clips, double bpm)
{
    std::map<std::string, std::vector<const UiProjectMidiClip*>> byNode;
    for (const auto& clip : clips)
        byNode[clip.targetNodeId].push_back(&clip);

    std::vector<std::vector<std::uint8_t>> tracks;
    tracks.push_back(buildTempoTrack(bpm));
    for (const auto& [nodeId, nodeclips] : byNode)
        tracks.push_back(buildNoteTrack(nodeId, nodeclips));

    const auto numTracks = static_cast<std::uint16_t>(tracks.size());
    std::vector<std::uint8_t> smf;
    smf.push_back('M'); smf.push_back('T');
    smf.push_back('h'); smf.push_back('d');
    writeU32BE(smf, 6);
    writeU16BE(smf, numTracks == 1 ? 0 : 1);
    writeU16BE(smf, numTracks);
    writeU16BE(smf, PPQ);
    for (const auto& track : tracks)
        smf.insert(smf.end(), track.begin(), track.end());
    return smf;
}

std::vector<std::uint8_t> buildCapturedTrack(
    const std::string& nodeId,
    const std::vector<const CapturedMidiEvent*>& events)
{
    const auto slash = nodeId.rfind('/');
    const auto label = slash != std::string::npos
        ? nodeId.substr(slash + 1) : nodeId;
    const auto nameLen = static_cast<std::uint8_t>(
        std::min(label.size(), std::size_t{127}));
    std::vector<RawEvent> trackEvents;
    std::vector<std::uint8_t> nameBytes = {0xFF, 0x03, nameLen};
    nameBytes.insert(nameBytes.end(), label.begin(), label.begin() + nameLen);
    trackEvents.push_back({0, std::move(nameBytes)});

    for (const auto* ev : events) {
        if ((ev->status & 0x80U) == 0) continue;
        trackEvents.push_back({beatToTick(ev->beatPosition),
            {ev->status, ev->data1, ev->data2}});
    }

    std::stable_sort(trackEvents.begin() + 1, trackEvents.end(),
        [](const RawEvent& a, const RawEvent& b) {
            return a.tick < b.tick;
        });

    const auto lastTick = trackEvents.size() > 1 ? trackEvents.back().tick : 0U;
    trackEvents.push_back({lastTick, {0xFF, 0x2F, 0x00}});
    return eventsToTrack(std::move(trackEvents));
}

std::string writeSmf(const std::string& path,
                     const std::vector<UiProjectMidiClip>& clips,
                     double bpm)
{
    const auto buffer = arrangementToSmf(clips, bpm);
    const auto parent = std::filesystem::path(path).parent_path();
    if (!parent.empty()) {
        std::error_code ec;
        std::filesystem::create_directories(parent, ec);
        if (ec) return "Unable to create directory: " + ec.message();
    }
    std::ofstream out(path, std::ios::binary | std::ios::trunc);
    if (!out) return "Unable to open output file: " + path;
    out.write(reinterpret_cast<const char*>(buffer.data()),
              static_cast<std::streamsize>(buffer.size()));
    if (!out) return "Failed to write MIDI file: " + path;
    return {};
}

std::string writeCapturedSmf(const std::string& path,
                             const std::vector<CapturedMidiEvent>& events,
                             double bpm)
{
    std::map<std::string, std::vector<const CapturedMidiEvent*>> byNode;
    for (const auto& ev : events)
        byNode[ev.nodeId].push_back(&ev);

    std::vector<std::vector<std::uint8_t>> tracks;
    tracks.push_back(buildTempoTrack(bpm));
    for (const auto& [nodeId, nodeEvents] : byNode)
        tracks.push_back(buildCapturedTrack(nodeId, nodeEvents));

    const auto numTracks = static_cast<std::uint16_t>(tracks.size());
    std::vector<std::uint8_t> smf;
    smf.push_back('M'); smf.push_back('T');
    smf.push_back('h'); smf.push_back('d');
    writeU32BE(smf, 6);
    writeU16BE(smf, numTracks == 1 ? 0 : 1);
    writeU16BE(smf, numTracks);
    writeU16BE(smf, PPQ);
    for (const auto& track : tracks)
        smf.insert(smf.end(), track.begin(), track.end());

    const auto parent = std::filesystem::path(path).parent_path();
    if (!parent.empty()) {
        std::error_code ec;
        std::filesystem::create_directories(parent, ec);
        if (ec) return "Unable to create directory: " + ec.message();
    }
    std::ofstream out(path, std::ios::binary | std::ios::trunc);
    if (!out) return "Unable to open output file: " + path;
    out.write(reinterpret_cast<const char*>(smf.data()),
              static_cast<std::streamsize>(smf.size()));
    if (!out) return "Failed to write MIDI file: " + path;
    return {};
}

} // namespace transmission
